### PR TITLE
completion fix

### PR DESCRIPTION
**Pull request by cossonleo**
Completion will not work correct When We type chars between  sended to completion request and recived the result.  These chars will not be replaced by completion items.

### CODE REVIEW

**Comment by pickfire** [comment]
How to reproduce the issue?

**Comment by archseer** [comment]
I don't quite understand the issue, we track the initial `trigger_offset` when `completion` is triggered: https://github.com/helix-editor/helix/blob/7eb013c6fb33df40b2fae31c9d08f7da275cd963/helix-term/src/commands.rs#L3490

Then we use that trigger offset when needed: https://github.com/helix-editor/helix/blob/7eb013c6fb33df40b2fae31c9d08f7da275cd963/helix-term/src/ui/completion.rs#L111-L117

We also remove any extra text: https://github.com/helix-editor/helix/blob/7eb013c6fb33df40b2fae31c9d08f7da275cd963/helix-term/src/ui/completion.rs#L133-L134

**Comment by archseer** [comment]
We might need to call `.savepoint` sooner, inside `fn completion` rather than here https://github.com/helix-editor/helix/blob/0062af6a19d96e0a4c3f94e3e44c179230e005b9/helix-term/src/ui/completion.rs#L152-L153

**Comment by cossonleo** [comment]
> We might need to call `.savepoint` sooner, inside `fn completion` rather than here
> 
> https://github.com/helix-editor/helix/blob/0062af6a19d96e0a4c3f94e3e44c179230e005b9/helix-term/src/ui/completion.rs#L152-L153

Yes
In the current implementation, there are still redundant characters.
It occurs when the completion request is sent and wait response. At this time, we type char again.

**Comment by cossonleo** [comment]
> How to reproduce the issue?

gopls 0.7.0
helix 0.6
.config/helix/config.toml
```
  idle-timeout = 40
```

helix_completion

**Comment by the-mikedavis** [comment]
I've been seeing the same completion problems with ElixirLS and ErlangLS. I'm not sure if it's the slowness of the LS or my low `idle-timeout` config but when I type and get completion as I type, `C-n`/`C-p`ing to one of the candidates does sometimes end up with a suffix like the screenshot. I hadn't really locked down what triggered it but it seems like if I deliberately pause before pressing `C-n`/`C-p` it doesn't happen (on master).

I just took this branch for a spin though and it seems to solve my completion problems.

**Comment by the-mikedavis** [comment]
I ran with this change yesterday and I noticed that the completion menu seems to pop up less. I imagine that what's happening is that I trigger the idle-timeout between keystrokes and the keystrokes following the auto-complete request cancel that request. (I haven't really looked at the change yet though so take that description with a grain of salt)

File: helix-view/src/document.rs [hunk 1]
```diff
@@ -112,7 +112,7 @@ pub struct Document {
     // be more troublesome.
     pub history: Cell<History>,
 
-    pub savepoint: Option<Transaction>,
+    pub savepoint: Option<(i32, Transaction)>,
```
Comment (pickfire) [inline helix-view/src/document.rs:115]: Why are we using `i32` for version instead of `u32`?

**Comment by archseer** [comment]
Hey @cossonleo I want to finally get this merged but I'm trying to understand the fix. Does it immediately set a save point, then when completion data arrives it filters that based on new input?

**Comment by cossonleo** [comment]
> Hey @cossonleo I want to finally get this merged but I'm trying to understand the fix. Does it immediately set a save point, then when completion data arrives it filters that based on new input?

Yes, Method `Completion::new()` will call `recompute_filter(editor)` to filter the data.

**Comment by cossonleo** [comment]
@archseer The PR is ready for merging. Do you have any other suggestions.

File: helix-view/src/document.rs (hunk 1 above)
Comment (cossonleo) [reply helix-view/src/document.rs:117]: The document  version type is `i32`.

**Comment by archseer** [comment]
Sorry, I'll look at this PR this week!

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -2884,12 +2882,37 @@ pub mod insert {
...
+            let mut iter = text.chars_at(cursor);
+            iter.reverse();
+            for _ in 0..config.completion_trigger_len {
+                if iter.next().map_or(true, |c| !char_is_word(c)) {
+                    return;
+                }
+            }
+            cx.editor.reset_idle_timer();
+            return;
+        }
```
Comment (pickfire) [inline helix-term/src/commands.rs:2912]: Completion happens only on the primary cursor? Does the same text gets shown for the rest of the selections?

File: helix-term/src/commands.rs [hunk 3]
```diff
@@ -3820,12 +3850,21 @@ pub fn completion(cx: &mut Context) {
...
+    doc.savepoint();
+    let trigger_version = doc.version();
     cx.callback(
         future,
         move |editor, compositor, response: Option<lsp::CompletionResponse>| {
-            let doc = doc!(editor);
+            let doc = doc_mut!(editor);
+            let savepoint = match doc.savepoint.take() {
+                Some(s) => s,
+                None => return,
```
Comment (pickfire) [inline helix-term/src/commands.rs:3861]: In what case this will happen? Can you please leave a comment?

File: helix-term/src/commands.rs [hunk 4]
```diff
@@ -2884,12 +2882,37 @@ pub mod insert {
             ..
         }) = &capabilities.completion_provider
         {
-            // TODO: what if trigger is multiple chars long
             if triggers.iter().any(|trigger| trigger.contains(ch)) {
-                cx.editor.clear_idle_timer();
-                super::completion(cx);
+                return true;
             }
```
Comment (pickfire) [inline helix-term/src/commands.rs:2887]:
```rust
            triggers.iter().any(|trigger| trigger.contains(ch))
```
Why not just returns this?

**Comment by archseer** [comment]
Looks like my comments didn't get posted last time: I think moving `doc.savepoint()` before the callback is the only change needed to fix the completion. Why is savepoint now storing version?

**Comment by cossonleo** [comment]
> Looks like my comments didn't get posted last time: I think moving `doc.savepoint()` before the callback is the only change needed to fix the completion. Why is savepoint now storing version?

We should discard the completion's result responsed after trigger the next completion. So we need  to store the version to determine whether to discard.

File: helix-term/src/commands.rs (hunk 3 above)
Comment (cossonleo) [reply helix-term/src/commands.rs:3861]: It's getted result after the mode was changed to normal and was changed to insert again.

File: helix-term/src/commands.rs (hunk 2 above)
Comment (cossonleo) [reply helix-term/src/commands.rs:2912]: I didn't consider the case. In this case, we should carefully consider what rules are needed

File: helix-term/src/commands.rs [hunk 5]
```diff
@@ -2903,11 +2901,35 @@ pub mod insert {
...
+        }
+    }
+
+    fn language_server_completion(cx: &mut Context, ch: char) {
+        use helix_core::chars::char_is_word;
+
+        let config = cx.editor.config();
+        if !config.auto_completion {
+            return;
+        }
```
Comment (pickfire) [inline helix-term/src/commands.rs:2916]: I think we need to add one more check to show completion beside this check, it's to check whether it's running from a macro (or `repeat_last_motion`), I notice this erratic behavior of completion popping up (and sometimes not closing) when I type <kbd>.</kbd>.

**Comment by sbromberger** [comment]
On my system, this patch causes completion to fail completely. Logs show
```
2022-10-06T17:19:39.010 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-10-06T17:19:39.012 helix_lsp::transport [ERROR] err: <- StreamClosed
```

This is gopls 0.9.5 with go 1.19.2.

**Comment by kirawi** [comment]
I think this is the code that handles it in VSCode: https://github.com/microsoft/vscode/blob/ba9399b83d048e780646896d5f81bcc08d759c6b/src/vs/editor/contrib/inlineCompletions/browser/inlineCompletionsModel.ts#L539-L548

**Comment by rcorre** [comment]
I don't have enough context to be confident in my conflict resolutions, but I merged this into `master` at https://github.com/rcorre/helix/commit/51f8b6732610b23011f38b2863afa26809b89bf0 and it seems to fix #4851 for me.

**Comment by pascalkuthe** [comment]
Superseded by  #6173

**Pull request closed without merging**